File: Learning Dynammic Programming/can_sum.py
from matplotlib.colors import cnames


def can_sum(target, nums, memo = {}):
    if (target in memo): return memo[target]
    if (target == 0): return True
    if (target < 0): return False

    for num in nums:
        if (can_sum(target - num, nums, memo) == True):
            memo[target] = True
            return True
    
    memo[target] = False
    return False

print(can_sum(300, [7,14]))

File: Learning Dynammic Programming/can_sum.cpp
#include<iostream>
#include<vector>
#include<unordered_map>
using namespace std;

bool can_sum(int targetSum, vector<int> num, unordered_map<int, bool> memo){
    if (memo.find(targetSum) != memo.end()) return memo[targetSum];

    if (targetSum == 0) return true;
    if (targetSum < 0) return false;

    for (int i = 0; i < num.size(); i++){
        if (can_sum(targetSum - num[i], num, memo) == true){
            memo[targetSum] = true;
            return true;
        }
    }

    memo[targetSum] = false;
    return false;
}

int main(){
    int s = 300;
    vector<int> n;
    unordered_map<int, bool> memo;

    n.push_back(7);
    n.push_back(14);
    // n.push_back(4);
    // n.push_back(7);

    cout<<boolalpha<<can_sum(s,n,memo)<<endl;

    return 0;
}